File: 2026-1/Graph-Algorithms/Tarea2-agra/20261-agra-hw02-samples/bit.cpp
/*
 *
 * B -> D: recursion que verifica si una region es toda igual.
 *         Si si, emite ese valor. Si no, emite 'D' y divide.
 *
 * D -> B: recursion que lee el siguiente caracter valido ('0','1','D')
 *         del input y rellena la grilla.
 *
 * COMPLEJIDAD: O(R*C) donde R=filas, C=columnas
 */

#include <bits/stdc++.h>
using namespace std;

char grilla[505][505];
string resultado;

// buffer para leer caracteres uno por uno
string bufer;
int pos;

// devuelve el proximo caracter valido ('0', '1', 'D')
char siguiente_char()
{
    while (true)
    {
        if (pos < (int)bufer.size())
        {
            char c = bufer[pos++];
            if (c == '0' || c == '1' || c == 'D')
                return c;
            // ignora saltos de linea y espacios
        }
        else
        {
            // leer mas caracteres del input
            char tmp[300005];
            if (scanf("%s", tmp) != 1)
                return 0;
            bufer = tmp;
            pos = 0;
        }
    }
}

// B -> D
void encode(int f, int c, int nf, int nc)
{
    char primero = grilla[f][c];
    bool todo_igual = true;
    for (int ff = f; ff < f + nf && todo_igual; ff++)
        for (int cc = c; cc < c + nc && todo_igual; cc++)
            if (grilla[ff][cc] != primero)
                todo_igual = false;

    if (todo_igual)
    {
        resultado += primero;
        return;
    }

    resultado += 'D';

    if (nf == 1)
    {
        int mitad_c = (nc + 1) / 2;
        encode(f, c, 1, mitad_c);
        encode(f, c + mitad_c, 1, nc - mitad_c);
    }
    else if (nc == 1)
    {
        int mitad_f = (nf + 1) / 2;
        encode(f, c, mitad_f, 1);
        encode(f + mitad_f, c, nf - mitad_f, 1);
    }
    else
    {
        int mitad_f = (nf + 1) / 2;
        int mitad_c = (nc + 1) / 2;
        encode(f, c, mitad_f, mitad_c);
        encode(f, c + mitad_c, mitad_f, nc - mitad_c);
        encode(f + mitad_f, c, nf - mitad_f, mitad_c);
        encode(f + mitad_f, c + mitad_c, nf - mitad_f, nc - mitad_c);
    }
}

// D -> B
void decode(int f, int c, int nf, int nc)
{
    char ch = siguiente_char();

    if (ch == '0' || ch == '1')
    {
        for (int ff = f; ff < f + nf; ff++)
            for (int cc = c; cc < c + nc; cc++)
                grilla[ff][cc] = ch;
        return;
    }

    // ch == 'D', dividir y seguir leyendo
    if (nf == 1)
    {
        int mitad_c = (nc + 1) / 2;
        decode(f, c, 1, mitad_c);
        decode(f, c + mitad_c, 1, nc - mitad_c);
    }
    else if (nc == 1)
    {
        int mitad_f = (nf + 1) / 2;
        decode(f, c, mitad_f, 1);
        decode(f + mitad_f, c, nf - mitad_f, 1);
    }
    else
    {
        int mitad_f = (nf + 1) / 2;
        int mitad_c = (nc + 1) / 2;
        decode(f, c, mitad_f, mitad_c);
        decode(f, c + mitad_c, mitad_f, nc - mitad_c);
        decode(f + mitad_f, c, nf - mitad_f, mitad_c);
        decode(f + mitad_f, c + mitad_c, nf - mitad_f, nc - mitad_c);
    }
}

// imprime la cadena en lineas de 50 caracteres
void imprimir(const string &s)
{
    for (int i = 0; i < (int)s.size(); i += 50)
        printf("%s\n", s.substr(i, 50).c_str());
}

int main()
{
    char fmt;

    while (scanf(" %c", &fmt) == 1 && fmt != '#')
    {
        int filas, cols;
        scanf("%d %d", &filas, &cols);

        // limpiar buffer para este caso :)
        bufer = "";
        pos = 0;

        if (fmt == 'B')
        {
            // leer filas*cols caracteres validos y llenar grilla
            int total = filas * cols;
            int leidos = 0;
            while (leidos < total)
            {
                char c = siguiente_char();
                grilla[leidos / cols][leidos % cols] = c;
                leidos++;
            }

            resultado = "";
            encode(0, 0, filas, cols);

            printf("D %4d %4d\n", filas, cols);
            imprimir(resultado);
        }
        else
        {
            // leer la cadena D caracter por caracter durante decode
            decode(0, 0, filas, cols);

            string salida = "";
            for (int f = 0; f < filas; f++)
                for (int c = 0; c < cols; c++)
                    salida += grilla[f][c];

            printf("B %4d %4d\n", filas, cols);
            imprimir(salida);
        }
    }

    return 0;
}

File: 2026-1/Graph-Algorithms/Tarea2-agra/20261-agra-hw02-samples/bit.py
#!/usr/bin/env python3
import sys

# grilla global para guardar los 0s y 1s
grilla = []
resultado = []

# buffer para leer caracter por caracter del input
char_buf = []
char_idx = 0

# devuelve el proximo caracter valido ('0', '1', 'D')
def siguiente_char():
    global char_buf, char_idx
    while True:
        if char_idx < len(char_buf):
            c = char_buf[char_idx]
            char_idx += 1
            if c in ('0', '1', 'D'):
                return c
            # ignora saltos de linea y espacios
        else:
            return None  # se acabo el input

# B -> D: si la region es toda igual emite ese valor, si no emite 'D' y divide
def encode(r, c, nr, nc):
    primero = grilla[r][c]
    todo_igual = True
    for rr in range(r, r + nr):
        for cc in range(c, c + nc):
            if grilla[rr][cc] != primero:
                todo_igual = False
                break
        if not todo_igual:
            break

    if todo_igual:
        resultado.append(primero)
        return

    resultado.append('D')

    if nr == 1:
        lc = (nc + 1) // 2
        encode(r, c, 1, lc)
        encode(r, c + lc, 1, nc - lc)
    elif nc == 1:
        tr = (nr + 1) // 2
        encode(r, c, tr, 1)
        encode(r + tr, c, nr - tr, 1)
    else:
        tr = (nr + 1) // 2
        lc = (nc + 1) // 2
        encode(r, c, tr, lc)
        encode(r, c + lc, tr, nc - lc)
        encode(r + tr, c, nr - tr, lc)
        encode(r + tr, c + lc, nr - tr, nc - lc)

# D -> B: lee el siguiente caracter y rellena la grilla
def decode(r, c, nr, nc):
    ch = siguiente_char()

    if ch in ('0', '1'):
        for rr in range(r, r + nr):
            for cc in range(c, c + nc):
                grilla[rr][cc] = ch
        return

    # ch == 'D', hay que dividir y seguir leyendo
    if nr == 1:
        lc = (nc + 1) // 2
        decode(r, c, 1, lc)
        decode(r, c + lc, 1, nc - lc)
    elif nc == 1:
        tr = (nr + 1) // 2
        decode(r, c, tr, 1)
        decode(r + tr, c, nr - tr, 1)
    else:
        tr = (nr + 1) // 2
        lc = (nc + 1) // 2
        decode(r, c, tr, lc)
        decode(r, c + lc, tr, nc - lc)
        decode(r + tr, c, nr - tr, lc)
        decode(r + tr, c + lc, nr - tr, nc - lc)

# imprime la cadena en lineas de 50 caracteres
def imprimir(s):
    for i in range(0, len(s), 50):
        print(s[i:i+50])

def main():
    global grilla, resultado, char_buf, char_idx

    datos = sys.stdin.read()
    char_buf = list(datos)
    char_idx = 0

    while True:
        # buscar el proximo formato valido ('B', 'D', o '#')
        fmt = None
        while char_idx < len(char_buf):
            c = char_buf[char_idx]
            char_idx += 1
            if c in ('B', 'D', '#'):
                fmt = c
                break

        if fmt is None or fmt == '#':
            break

        # leer filas y columnas
        nums = []
        num_actual = ""
        while len(nums) < 2:
            if char_idx >= len(char_buf):
                break
            c = char_buf[char_idx]
            char_idx += 1
            if c.isdigit():
                num_actual += c
            elif num_actual:
                nums.append(int(num_actual))
                num_actual = ""
        if num_actual:
            nums.append(int(num_actual))

        filas, cols = nums[0], nums[1]

        # reiniciar grilla para este caso
        grilla = [['' for _ in range(cols)] for _ in range(filas)]

        if fmt == 'B':
            # leer filas*cols caracteres validos y llenar grilla
            leidos = 0
            total = filas * cols
            while leidos < total:
                c = siguiente_char()
                grilla[leidos // cols][leidos % cols] = c
                leidos += 1

            resultado = []
            encode(0, 0, filas, cols)

            print(f"D {filas:4d} {cols:4d}")
            imprimir("".join(resultado))

        else:
            # leer la cadena D caracter por caracter durante el decode
            decode(0, 0, filas, cols)

            salida = ""
            for r in range(filas):
                for c in range(cols):
                    salida += grilla[r][c]

            print(f"B {filas:4d} {cols:4d}")
            imprimir(salida)

if __name__ == "__main__":
    # Python tiene limite de recursion, lo subimos por si acaso
    sys.setrecursionlimit(300000)
    main()